docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,381 @@
+│   ├── meta_test.go
+│   ├── restmapper.go
+│   └── restmapper_test.go
+├── meta.go
+├── meta_test.go
+├── node_example.json
+├── pod_example.json
+├── ref.go
+├── ref_test.go
+├── registered
+│   └── registered.go
+├── register.go
+├── replication_controller_example.json
+├── requestcontext.go
+├── resource
+│   ├── quantity_example_test.go
+│   ├── quantity.go
+│   ├── quantity_test.go
+│   └── suffix.go
+├── resource_helpers.go
+├── resource_helpers_test.go
+├── rest
+│   ├── create.go
+│   ├── create_test.go
+│   ├── delete.go
+│   ├── doc.go
+│   ├── rest.go
+│   ├── resttest
+│   │   └── resttest.go
+│   ├── types.go
+│   ├── update.go
+│   └── update_test.go
+├── serialization_test.go
+├── testapi
+│   ├── testapi.go
+│   └── testapi_test.go
+├── testing
+│   ├── compat
+│   │   └── compatibility_tester.go
+│   ├── fuzzer.go
+│   └── pod_specs.go
+├── types.generated.go
+├── types.go
+├── unversioned
+│   ├── duration.go
+│   ├── duration_test.go
+│   ├── time.go
+│   ├── time_test.go
+│   ├── types.go
+│   └── types_swagger_doc_generated.go
+├── util
+│   ├── group_version.go
+│   └── group_version_test.go
+├── v1
+│   ├── backward_compatibility_test.go
+│   ├── conversion_generated.go
+│   ├── conversion.go
+│   ├── conversion_test.go
+│   ├── deep_copy_generated.go
+│   ├── defaults.go
+│   ├── defaults_test.go
+│   ├── doc.go
+│   ├── register.go
+│   ├── types.generated.go
+│   ├── types.go
+│   └── types_swagger_doc_generated.go
+└── validation
+    ├── doc.go
+    ├── events.go
+    ├── events_test.go
+    ├── schema.go
+    ├── schema_test.go
+    ├── testdata
+    │   └── v1
+    │       ├── invalidPod1.json
+    │       ├── invalidPod2.json
+    │       ├── invalidPod3.json
+    │       ├── invalidPod.yaml
+    │       └── validPod.yaml
+    ├── validation.go
+    └── validation_test.go
+
+```
+
+The api in `pkg/apis` is closer to where we want to be.
+
+```
+pkg/apis/
+└── extensions
+    ├── deep_copy_generated.go
+    ├── helpers.go
+    ├── helpers_test.go
+    ├── install
+    │   ├── install.go
+    │   └── install_test.go
+    ├── register.go
+    ├── types.generated.go
+    ├── types.go
+    ├── v1beta1
+    │   ├── conversion_generated.go
+    │   ├── conversion.go
+    │   ├── deep_copy_generated.go
+    │   ├── defaults.go
+    │   ├── defaults_test.go
+    │   ├── register.go
+    │   ├── types.generated.go
+    │   ├── types.go
+    │   └── types_swagger_doc_generated.go
+    └── validation
+        ├── validation.go
+        └── validation_test.go
+```
+
+## Desired directory structure
+
+We want to define a standard layout and structure for Kubernetes API types.
+Below is a first draft, with explanation afterwards. We may try to move the
+relevant parts of the legacy `pkg/api/v1` api into the new structure, although
+this api doesn't have a group.
+
+```
+apis
+└── group
+    ├── install
+    ├── types.go
+    ├── v1
+    │   ├── install
+    │   │   ├── conversion
+    │   │   │   ├── generated_conversion.go
```

Comment:
Yeah, I left encoding generation out of this, but I'd like encoding/decoding generator(s) to follow the same package structure. We can write our own if necessary. I assume we will write our own when we want to do binary encodings.